llm/cpp/llm.cpp

```diff
@@ -51,38 +202,209 @@ int main(int argc, char* argv[]) try {
+            for (size_t cand_id = 0; cand_id < candidates.size(); ++cand_id) {
+                if (EOS_TOKEN == candidates[cand_id].tokens.back()) {  // TODO: idx->token_id
+                    // if beam_token does not belong to top num_beams tokens, it should not be added
+                    if (cand_id >= GROUP_SIZE) {
+                        continue;
+                    }
+                    candidates[cand_id].tokens.resize(candidates[cand_id].tokens.size() - 1);
+                    groups[group_idx].hypotheses.push(std::move(candidates[cand_id]), prompt_length);
+                } else {
+                    // if (candidates[cand_id].tokens.size() > 1 && candidates[cand_id].tokens[candidates[cand_id].tokens.size() - 2] == 4030) {
+                    //     std::cout << candidates[cand_id].tokens.back() << ' ' << candidates[cand_id].log_prob << '\n';
+                    // }
+                    // std::cout << candidates[cand_id].log_prob << ", ";
+                    groups[group_idx].beams.push_back(std::move(candidates[cand_id]));
+                    size_t cur_beam = groups[group_idx].beams.size() - 1;  // TODO: beter loop iteration
+                    auto ireq = compiled.create_infer_request();
+                    for (size_t tensor_id = 3; tensor_id < inputs.size(); ++tensor_id) {
+                        ireq.set_input_tensor(tensor_id, groups[group_idx].beams.back().ireq.get_output_tensor(tensor_id - 2));
+                    }
+                    ireq.get_tensor("input_ids").set_shape({BATCH_SIZE, 1});
+                    ireq.get_tensor("input_ids").data<int64_t>()[0] = groups[group_idx].beams.back().tokens.back();
+                    ireq.get_tensor("attention_mask").set_shape({BATCH_SIZE, groups[group_idx].beams.back().ireq.get_tensor("attention_mask").get_size() + 1});
+                    std::fill_n(ireq.get_tensor("attention_mask").data<int64_t>(), ireq.get_tensor("attention_mask").get_size(), 1);
+                    ireq.get_tensor("position_ids").set_shape({BATCH_SIZE, 1});
+                    ireq.get_tensor("position_ids").data<int64_t>()[0] = ireq.get_tensor("attention_mask").get_size() - 1;
+                    groups[group_idx].beams.back().ireq = ireq;
+                    if (groups[group_idx].beams.size() == GROUP_SIZE) {
+                        break;
+                    }
+                }
+            }
+            groups[group_idx].hypotheses.is_done(cur_len + prompt_length, groups[group_idx].beams.front().log_prob);  // TODO: that requires groups[group_idx].beams to be not empty
+            // if (std::all_of(groups.begin(), groups.end(), [cur_len, prompt_length](Group& gr){return gr.hypotheses.is_done(cur_len + prompt_length, gr.beams.front().log_prob);})) {  // TODO: that requires groups[group_idx].beams to be not empty
+            //     break;
+            // }
+
+            for (Beam& beam : groups[group_idx].beams) {
+                beam.ireq.infer();
```

Comment:
This is a sync call, so we are processing all beams sequentially.

Reply:
True. After I finalize the alignment with transformers there will by async. Replace `beam.ireq.infer();` with `beam.ireq.start_async();`, replace `// beam_ireq.wait();  TODO: async` with `beam_ireq.wait();`.

Reply:
Removed